compiler/rustc_target/src/asm/mips.rs

```diff
@@ -0,0 +1,106 @@
+
+def_reg_class! {
+    Mips MipsInlineAsmRegClass {
+        reg,
+    }
+}
+
+impl MipsInlineAsmRegClass {
+    pub fn valid_modifiers(self, _arch: super::InlineAsmArch) -> &'static [char] {
+        &[]
+    }
+
+    pub fn suggest_class(self, _arch: InlineAsmArch, _ty: InlineAsmType) -> Option<Self> {
+        None
+    }
+
+    pub fn suggest_modifier(
+        self,
+        _arch: InlineAsmArch,
+        _ty: InlineAsmType,
+    ) -> Option<(char, &'static str)> {
+        None
+    }
+
+    pub fn default_modifier(self, _arch: InlineAsmArch) -> Option<(char, &'static str)> {
+        None
+    }
+
+    pub fn supported_types(
+        self,
+        _arch: InlineAsmArch,
+    ) -> &'static [(InlineAsmType, Option<&'static str>)] {
+        match self {
+            Self::reg => types! { _: I8, I16, I32; },
+        }
+    }
+}
+
+fn eliminate_frame_pointer(
+    _arch: InlineAsmArch,
+    _has_feature: impl FnMut(&str) -> bool,
+    target: &Target,
+    _allocating: bool,
+) -> Result<(), &'static str> {
+    if target.options.eliminate_frame_pointer {
+        Ok(())
+    } else {
+        Err(
+            "frame pointer cannot be used as an operand when disabling `eliminate_frame_pointer` feature",
+        )
+    }
+}
+
+// The reserved registers are somewhat taken from <https://git.io/JUR1k#L150>.
+def_regs! {
+    Mips MipsInlineAsmReg MipsInlineAsmRegClass {
+        v0: reg = ["v0", "2"],
+        v1: reg = ["v1", "3"],
+        a0: reg = ["a0", "4"],
+        a1: reg = ["a1", "5"],
+        a2: reg = ["a2", "6"],
+        a3: reg = ["a3", "7"],
+        t0: reg = ["t0", "8"],
+        t1: reg = ["t1", "9"],
+        t2: reg = ["t2", "10"],
+        t3: reg = ["t3", "11"],
+        t4: reg = ["t4", "12"],
+        t5: reg = ["t5", "13"],
+        t6: reg = ["t6", "14"],
+        t7: reg = ["t7", "15"],
+        s0: reg = ["s0", "16"],
+        s1: reg = ["s1", "17"],
+        s2: reg = ["s2", "18"],
+        s3: reg = ["s3", "19"],
+        s4: reg = ["s4", "20"],
+        s5: reg = ["s5", "21"],
+        s6: reg = ["s6", "22"],
+        s7: reg = ["s7", "23"],
+        t8: reg = ["t8", "24"],
+        t9: reg = ["t9", "25"],
+        fp: reg = ["s8", "fp", "30"] % eliminate_frame_pointer,
+        #error = ["zero", "0"] => "constant zero cannot be used as an operand for inline asm",
+        #error = ["at", "1"] => "reserved for assembler (Assembler Temp)",
+        #error = ["k0", "26"] => "OS-reserved register cannot be used as an operand for inline asm",
+        #error = ["k1", "27"] =>  "OS-reserved register cannot be used as an operand for inline asm",
+        #error = ["gp", "28"] => "the global pointer cannot be used as an operand for inline asm",
+        #error = ["sp", "29"] => "the stack pointer cannot be used as an operand for inline asm",
+        // FIXME: Reserve $ra, $t0, $t1 if in mips16 mode.
+        // #error = ["ra", "31"] => "the return address register cannot be used as an operand for inline asm",
```

Comment:
To avoid backwards compatibility issues, could you unconditionally reserve them for now and make the `FIXME` for re-enabling when not in 16-bit mode?

Reply:
I reserve $ra for now. I leave $t0 and $t1 intact because Rust hasn't supported mips16 yet.